File: archive/voice-analysis/src/processing/match_pairer.py
"""Pair QW Hub matches to Craig recording window. Calculate audio offsets and confidence."""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from datetime import datetime, timezone

from src.processing.craig_parser import CraigSession

logger = logging.getLogger(__name__)


@dataclass
class MatchPairing:
    """A QW Hub match paired to a position in the Craig recording."""

    match_id: int
    map_name: str
    timestamp: datetime
    server_hostname: str
    teams: list[dict] = field(default_factory=list)
    players: list[dict] = field(default_factory=list)
    ktxstats: dict | None = None
    duration_seconds: float = 1200.0
    audio_offset_seconds: float = 0.0
    audio_end_seconds: float = 0.0
    confidence: float = 0.0
    confidence_reasons: list[str] = field(default_factory=list)
    demo_sha256: str = ""


def pair_matches(
    craig_session: CraigSession,
    hub_matches: list[dict],
    ktxstats_map: dict[str, dict] | None = None,
    clock_tolerance_seconds: float = 5.0,
    padding_seconds: float = 10.0,
    default_duration: float = 1200.0,
) -> list[MatchPairing]:
    """Pair QW Hub matches to positions in the Craig recording.

    Args:
        craig_session: Parsed Craig session with start_time.
        hub_matches: List of match dicts from QW Hub API.
        ktxstats_map: Dict mapping demo_sha256 -> ktxstats dict.
        clock_tolerance_seconds: Allowed clock drift between Craig and QW Hub.
        padding_seconds: Extra seconds before/after match for audio slice.
        default_duration: Default match duration if ktxstats unavailable.

    Returns:
        List of MatchPairing objects sorted by audio offset.
    """
    ktxstats_map = ktxstats_map or {}
    pairings = []

    for match in hub_matches:
        match_ts_str = match.get("timestamp", "")
        if not match_ts_str:
            logger.warning("Match %s has no timestamp, skipping", match.get("id"))
            continue

        match_ts = datetime.fromisoformat(match_ts_str)
        if match_ts.tzinfo is None:
            match_ts = match_ts.replace(tzinfo=timezone.utc)

        craig_start = craig_session.start_time
        if craig_start.tzinfo is None:
            craig_start = craig_start.replace(tzinfo=timezone.utc)

        # Calculate audio offset: how many seconds into the recording this match starts
        offset = (match_ts - craig_start).total_seconds()

        # Get ktxstats if available
        demo_sha = match.get("demo_sha256", "")
        ktxstats = ktxstats_map.get(demo_sha)
        duration = default_duration

        if ktxstats and ktxstats.get("duration"):
            duration = float(ktxstats["duration"])

        audio_start = offset - padding_seconds
        audio_end = offset + duration + padding_seconds

        # Score confidence
        confidence, reasons = _score_confidence(
            match, craig_session, offset, ktxstats is not None
        )

        pairing = MatchPairing(
            match_id=match.get("id", 0),
            map_name=match.get("map", "unknown"),
            timestamp=match_ts,
            server_hostname=match.get("hostname", ""),
            teams=match.get("teams", []),
            players=match.get("players", []),
            ktxstats=ktxstats,
            duration_seconds=duration,
            audio_offset_seconds=audio_start,
            audio_end_seconds=audio_end,
            confidence=confidence,
            confidence_reasons=reasons,
            demo_sha256=demo_sha,
        )
        pairings.append(pairing)

    # Sort by audio offset
    pairings.sort(key=lambda p: p.audio_offset_seconds)

    # Validate no overlapping segments
    pairings = _validate_no_overlap(pairings)

    for p in pairings:
        logger.info(
            "Paired match %d (%s) at offset %.1fs-%.1fs, confidence=%.2f",
            p.match_id, p.map_name, p.audio_offset_seconds, p.audio_end_seconds, p.confidence,
        )

    return pairings


def _score_confidence(
    match: dict,
    craig_session: CraigSession,
    audio_offset: float,
    has_ktxstats: bool,
) -> tuple[float, list[str]]:
    """Score confidence of a match pairing.

    Returns:
        Tuple of (confidence_score 0.0-1.0, list of reason strings).
    """
    score = 0.0
    reasons = []

    # Factor 1: Offset is positive and reasonable (weight 0.3)
    if audio_offset > 60.0:
        score += 0.3
        reasons.append("offset > 60s into recording")
    elif audio_offset > 0:
        score += 0.15
        reasons.append(f"offset positive but small ({audio_offset:.1f}s)")
    else:
        reasons.append(f"offset negative ({audio_offset:.1f}s) - match before recording")

    # Factor 2: ktxstats available (weight 0.2)
    if has_ktxstats:
        score += 0.2
        reasons.append("ktxstats available (exact duration)")
    else:
        reasons.append("no ktxstats (using default duration)")

    # Factor 3: Player name overlap (weight 0.3)
    craig_names = set()
    for track in craig_session.tracks:
        craig_names.add(track.discord_username.lower())
        craig_names.add(track.discord_display_name.lower())

    match_names = set()
    for player in match.get("players", []):
        match_names.add(player.get("name", "").lower())

    overlap = craig_names & match_names
    if len(overlap) >= 3:
        score += 0.3
        reasons.append(f"{len(overlap)} player names match Craig tracks")
    elif len(overlap) >= 1:
        frac = len(overlap) / 3.0
        score += 0.3 * frac
        reasons.append(f"{len(overlap)} player name(s) match Craig tracks")
    else:
        reasons.append("no player name overlap with Craig tracks")

    # Factor 4: Offset within recording duration (weight 0.2)
    # Estimate recording duration from longest track (if we had it)
    # For now, check that offset is within a reasonable session window (4 hours)
    max_session = 4 * 3600  # 4 hours
    if 0 < audio_offset < max_session:
        score += 0.2
        reasons.append("offset within reasonable session window")
    else:
        reasons.append(f"offset {audio_offset:.1f}s outside expected range")

    return round(score, 2), reasons


def _validate_no_overlap(pairings: list[MatchPairing]) -> list[MatchPairing]:
    """Validate that pairings don't overlap and trim if needed."""
    if len(pairings) < 2:
        return pairings

    validated = [pairings[0]]
    for i in range(1, len(pairings)):
        prev = validated[-1]
        curr = pairings[i]

        if curr.audio_offset_seconds < prev.audio_end_seconds:
            # Overlap detected - trim the padding
            midpoint = (prev.audio_end_seconds + curr.audio_offset_seconds) / 2
            prev.audio_end_seconds = midpoint
            curr.audio_offset_seconds = midpoint
            logger.warning(
                "Trimmed overlap between match %d and %d at %.1fs",
                prev.match_id, curr.match_id, midpoint,
            )

        validated.append(curr)

    return validated


def format_pairing_summary(pairings: list[MatchPairing]) -> str:
    """Format a human-readable summary of match pairings."""
    if not pairings:
        return "No matches paired to this recording."

    lines = [f"Found {len(pairings)} match(es):\n"]
    for i, p in enumerate(pairings, 1):
        team_str = " vs ".join(
            f"{t.get('name', '?')} ({t.get('frags', '?')})"
            for t in p.teams
        )
        lines.append(f"  [{i}] {p.map_name} - {team_str}")
        lines.append(f"      Time: {p.timestamp.strftime('%H:%M:%S')} UTC")
        lines.append(f"      Audio: {p.audio_offset_seconds:.1f}s -> {p.audio_end_seconds:.1f}s")
        lines.append(f"      Duration: {p.duration_seconds:.0f}s")
        lines.append(f"      Confidence: {p.confidence:.0%}")
        for reason in p.confidence_reasons:
            lines.append(f"        - {reason}")
        lines.append("")

    return "\n".join(lines)


if __name__ == "__main__":
    # Quick test with hardcoded values from PLAN.md
    from datetime import datetime, timezone

    logging.basicConfig(level=logging.INFO, format="%(levelname)s: %(message)s")

    # Simulated Craig session
    session = CraigSession(
        start_time=datetime(2026, 2, 1, 15, 54, 44, 674000, tzinfo=timezone.utc),
        tracks=[],
    )

    # Simulated match
    test_match = {
        "id": 12345,
        "timestamp": "2026-02-01T15:55:36+00:00",
        "map": "dm4",
        "demo_sha256": "abc123",
        "teams": [{"name": "]sr[", "frags": 100}, {"name": "red", "frags": 80}],
        "players": [{"name": "ParadokS"}, {"name": "Razor"}],
        "hostname": "Test Server",
    }

    pairings = pair_matches(session, [test_match])
    print(format_pairing_summary(pairings))
    # Expected offset: ~51.326s (15:55:36 - 15:54:44.674)

File: apps/quad/src/modules/processing/stages/match-pairer.ts
/**
 * Match pairer — pairs QW Hub matches to positions in a recording session.
 * Calculates audio offsets, confidence scores, and validates no overlaps.
 *
 * Ported from voice-analysis/src/processing/match_pairer.py
 */

import { logger } from '../../../core/logger.js';
import type { HubMatch, HubTeam, KtxStats, MatchPairing, SessionMetadata } from '../types.js';

const DEFAULT_MATCH_LENGTH = 1210; // 10s countdown + 20min gameplay
const MAX_SESSION_HOURS = 4;
const MIN_PLAYER_OVERLAP = 2; // Minimum QW player names that must match for a "mix" game

/** Strip QW decorations and Discord punctuation for name comparison.
 *  Hub names have bullets ("• razor"), team prefixes, etc.
 *  Discord names may have dots (".andeh"), underscores, brackets. */
function normalizeName(name: string): string {
  return name.replace(/^[^a-zA-Z0-9]+/, '').replace(/[^a-zA-Z0-9]+$/, '').toLowerCase();
}

export interface PairMatchesOptions {
  defaultDuration?: number;
  /** Team tag to filter matches (e.g., "]sr["). Primary filter. */
  teamTag?: string;
  /** Discord ID → QW name mapping from bot registration. Used for player-based filtering. */
  knownPlayers?: Record<string, string>;
}

/**
 * Pair QW Hub matches to positions in the recording.
 *
 * Filtering strategy:
 * 1. If teamTag or knownPlayers are provided, only keep matches where:
 *    - A team name matches the team tag (case-insensitive), OR
 *    - At least MIN_PLAYER_OVERLAP resolved QW names appear in the match players
 * 2. Matches that pass neither filter are discarded (not our matches).
 * 3. Remaining matches get precise audio offsets from timestamps.
 *
 * @param session Recording session metadata
 * @param hubMatches Matches from QW Hub API
 * @param ktxstatsMap Map of demo_sha256 -> KtxStats
 * @param options Pairing options (team tag, known players, default duration)
 * @returns Pairings sorted by audio offset, trimmed for overlaps
 */
export function pairMatches(
  session: SessionMetadata,
  hubMatches: HubMatch[],
  ktxstatsMap?: Map<string, KtxStats>,
  options?: PairMatchesOptions,
): MatchPairing[] {
  const defaultMatchLength = options?.defaultDuration ?? DEFAULT_MATCH_LENGTH;
  const statsMap = ktxstatsMap ?? new Map<string, KtxStats>();
  const teamTag = options?.teamTag?.toLowerCase() ?? '';
  const knownPlayers = options?.knownPlayers ?? {};

  // Resolve recording tracks to QW names.
  // Mumble source: username IS the QW name (registered via ICE with QW display name).
  // Discord source: look up via knownPlayers[discordUserId] mapping.
  const isMumbleSource = session.source === 'mumble';
  const resolvedQwNames = new Set<string>();
  for (const track of session.tracks) {
    if (isMumbleSource && track.mumble_username) {
      resolvedQwNames.add(track.mumble_username.toLowerCase());
    } else if (track.discord_user_id) {
      const qwName = knownPlayers[track.discord_user_id];
      if (qwName) {
        resolvedQwNames.add(qwName.toLowerCase());
      }
    }
  }

  if (teamTag || resolvedQwNames.size > 0) {
    logger.info('Match filtering enabled', {
      teamTag: teamTag || '(none)',
      resolvedPlayers: [...resolvedQwNames],
    });
  } else {
    logger.warn('No team tag or player mapping — all hub matches will be paired (unfiltered)');
  }

  // Filter hub matches to only our team's matches
  const hasFilters = teamTag !== '' || resolvedQwNames.size > 0;
  const filtered = hasFilters ? filterOurMatches(hubMatches, teamTag, resolvedQwNames) : hubMatches;

  logger.info('Hub matches filtered', {
    total: hubMatches.length,
    kept: filtered.length,
    discarded: hubMatches.length - filtered.length,
  });

  const recordingStart = new Date(session.recording_start_time);
  const pairings: MatchPairing[] = [];

  for (const match of filtered) {
    const matchTsStr = match.timestamp;
    if (!matchTsStr) {
      logger.warn('Match has no timestamp, skipping', { matchId: match.id });
      continue;
    }

    const matchTs = new Date(matchTsStr);

    // Get ktxstats if available
    const demoSha = match.demo_sha256 ?? '';
    const ktxstats = statsMap.get(demoSha) ?? null;

    // Hub timestamp = demo start (= countdown start, derived from MVD filename).
    // ktxstats.date = match end time, ktxstats.duration = gameplay only (excl countdown).
    // Audio should span from demo start to match end (countdown + gameplay).
    const audioStart = (matchTs.getTime() - recordingStart.getTime()) / 1000;

    let audioEnd: number;
    let duration: number;
    const ktxDateStr = ktxstats?.date as string | undefined;
    if (ktxDateStr) {
      // Exact end time from ktxstats
      // Format: "2026-02-17 20:58:51 +0100" → "2026-02-17T20:58:51+01:00"
      const ktxEnd = new Date(
        ktxDateStr
          .replace(' ', 'T')                              // date-time separator
          .replace(/ ([+-])(\d{2})(\d{2})$/, '$1$2:$3'),  // " +0100" → "+01:00"
      );
      audioEnd = (ktxEnd.getTime() - recordingStart.getTime()) / 1000;
      duration = audioEnd - audioStart;

      // NaN guard — if ktxstats date didn't parse, fall back to default duration
      if (!Number.isFinite(audioEnd) || !Number.isFinite(duration) || duration <= 0) {
        logger.warn('Invalid ktxstats date, using default duration', {
          matchId: match.id,
          ktxDateStr,
          audioEnd,
          duration,
        });
        duration = defaultMatchLength;
        audioEnd = audioStart + duration;
      }
    } else {
      duration = defaultMatchLength;
      audioEnd = audioStart + duration;
    }

    // Score confidence
    const { score, reasons } = scoreConfidence(match, session, audioStart, ktxstats !== null, knownPlayers);

    pairings.push({
      matchId: match.id ?? 0,
      mapName: match.map ?? 'unknown',
      timestamp: matchTs,
      serverHostname: match.hostname ?? '',
      teams: match.teams ?? [],
      players: match.players ?? [],
      ktxstats,
      durationSeconds: duration,
      audioOffsetSeconds: audioStart,
      audioEndSeconds: audioEnd,
      confidence: score,
      confidenceReasons: reasons,
      demoSha256: demoSha,
    });
  }

  // Sort by audio offset
  pairings.sort((a, b) => a.audioOffsetSeconds - b.audioOffsetSeconds);

  // Validate no overlapping segments
  validateNoOverlap(pairings);

  for (const p of pairings) {
    logger.info('Paired match', {
      matchId: p.matchId,
      map: p.mapName,
      offsetStart: p.audioOffsetSeconds.toFixed(1),
      offsetEnd: p.audioEndSeconds.toFixed(1),
      confidence: p.confidence,
    });
  }

  return pairings;
}

/**
 * Filter hub matches to only those belonging to our team.
 *
 * A match passes if:
 * - Any team name matches the team tag (case-insensitive), OR
 * - At least MIN_PLAYER_OVERLAP of the resolved QW names appear in the match
 */
function filterOurMatches(
  hubMatches: HubMatch[],
  teamTag: string,
  resolvedQwNames: Set<string>,
): HubMatch[] {
  return hubMatches.filter((match) => {
    // Check team tag match
    if (teamTag) {
      const tagMatch = (match.teams ?? []).some(
        (t) => t.name?.toLowerCase() === teamTag,
      );
      if (tagMatch) return true;
    }

    // Check player name overlap (normalized — strips QW decorations + Discord punctuation)
    if (resolvedQwNames.size > 0) {
      const matchPlayerNorms = (match.players ?? [])
        .map((p) => p.name ? normalizeName(p.name) : '')
        .filter(Boolean);
      let overlap = 0;
      for (const qwName of resolvedQwNames) {
        const norm = normalizeName(qwName);
        if (norm.length < 2) continue;
        if (matchPlayerNorms.some((hubNorm) => hubNorm.includes(norm) || norm.includes(hubNorm))) overlap++;
      }
      if (overlap >= MIN_PLAYER_OVERLAP) return true;
    }

    return false;
  });
}

/**
 * Score confidence of a match pairing (0.0 - 1.0).
 *
 * Factors:
 *   - Offset is positive and reasonable (weight 0.3)
 *   - ktxstats available for exact duration (weight 0.2)
 *   - Player name overlap between recording tracks and match players (weight 0.3)
 *   - Offset within session window (weight 0.2)
 */
function scoreConfidence(
  match: HubMatch,
  session: SessionMetadata,
  audioOffset: number,
  hasKtxstats: boolean,
  knownPlayers: Record<string, string> = {},
): { score: number; reasons: string[] } {
  let score = 0;
  const reasons: string[] = [];

  // Factor 1: Offset is positive and reasonable (weight 0.3)
  if (audioOffset > 60) {
    score += 0.3;
    reasons.push('offset > 60s into recording');
  } else if (audioOffset > 0) {
    score += 0.15;
    reasons.push(`offset positive but small (${audioOffset.toFixed(1)}s)`);
  } else {
    reasons.push(`offset negative (${audioOffset.toFixed(1)}s) - match before recording`);
  }

  // Factor 2: ktxstats available (weight 0.2)
  if (hasKtxstats) {
    score += 0.2;
    reasons.push('ktxstats available (exact duration)');
  } else {
    reasons.push('no ktxstats (using default duration)');
  }

  // Factor 3: Player name overlap (weight 0.3)
  // Hub player names have QW decorations (e.g., "• razor", "tco.........axe")
  // Use substring matching: check if hub name contains any known name
  const sessionNames = new Set<string>();
  const isMumble = session.source === 'mumble';
  for (const track of session.tracks) {
    if (isMumble && track.mumble_username) {
      // Mumble: username IS the QW name
      sessionNames.add(normalizeName(track.mumble_username));
    } else {
      // Discord: resolved QW names from knownPlayers (most reliable — stable nicks)
      if (track.discord_user_id) {
        const qwName = knownPlayers[track.discord_user_id];
        if (qwName) sessionNames.add(normalizeName(qwName));
      }
      // Discord names as fallback
      if (track.discord_username) sessionNames.add(normalizeName(track.discord_username));
      if (track.discord_display_name) sessionNames.add(normalizeName(track.discord_display_name));
    }
  }

  const matchPlayerNorms = (match.players ?? [])
    .map((p) => p.name ? normalizeName(p.name) : '')
    .filter(Boolean);

  let overlapCount = 0;
  for (const name of sessionNames) {
    if (name.length < 3) continue; // Skip very short names to avoid false positives
    if (matchPlayerNorms.some((hubNorm) => hubNorm.includes(name) || name.includes(hubNorm))) overlapCount++;
  }

  if (overlapCount >= 3) {
    score += 0.3;
    reasons.push(`${overlapCount} player names match recording tracks`);
  } else if (overlapCount >= 1) {
    const frac = overlapCount / 3;
    score += 0.3 * frac;
    reasons.push(`${overlapCount} player name(s) match recording tracks`);
  } else {
    reasons.push('no player name overlap with recording tracks');
  }

  // Factor 4: Offset within recording duration (weight 0.2)
  const maxSession = MAX_SESSION_HOURS * 3600;
  if (audioOffset > 0 && audioOffset < maxSession) {
    score += 0.2;
    reasons.push('offset within reasonable session window');
  } else {
    reasons.push(`offset ${audioOffset.toFixed(1)}s outside expected range`);
  }

  return { score: Math.round(score * 100) / 100, reasons };
}

/**
 * Trim overlapping pairings at their midpoint (mutates in place).
 */
function validateNoOverlap(pairings: MatchPairing[]): void {
  for (let i = 1; i < pairings.length; i++) {
    const prev = pairings[i - 1];
    const curr = pairings[i];

    if (curr.audioOffsetSeconds < prev.audioEndSeconds) {
      const midpoint = (prev.audioEndSeconds + curr.audioOffsetSeconds) / 2;
      prev.audioEndSeconds = midpoint;
      curr.audioOffsetSeconds = midpoint;
      logger.warn('Trimmed overlap between matches', {
        matchA: prev.matchId,
        matchB: curr.matchId,
        midpoint: midpoint.toFixed(1),
      });
    }
  }
}

/**
 * Format a human-readable summary of match pairings.
 */
export function formatPairingSummary(pairings: MatchPairing[]): string {
  if (pairings.length === 0) {
    return 'No matches paired to this recording.';
  }

  const lines: string[] = [`Found ${pairings.length} match(es):\n`];

  for (let i = 0; i < pairings.length; i++) {
    const p = pairings[i];
    const teamStr = p.teams
      .map((t: HubTeam) => `${t.name ?? '?'} (${t.frags ?? '?'})`)
      .join(' vs ');

    lines.push(`  [${i + 1}] ${p.mapName} - ${teamStr}`);
    lines.push(`      Time: ${p.timestamp.toISOString().slice(11, 19)} UTC`);
    lines.push(`      Audio: ${p.audioOffsetSeconds.toFixed(1)}s -> ${p.audioEndSeconds.toFixed(1)}s`);
    lines.push(`      Duration: ${p.durationSeconds.toFixed(0)}s`);
    lines.push(`      Confidence: ${(p.confidence * 100).toFixed(0)}%`);
    for (const reason of p.confidenceReasons) {
      lines.push(`        - ${reason}`);
    }
    lines.push('');
  }

  return lines.join('\n');
}
